h2o-genmodel/src/main/java/hex/genmodel/MojoReaderBackendFactory.java

```diff
@@ -0,0 +1,116 @@
+ *
+ * <pre>
+ * {@code
+ *   public class ExampleApp {
+ *     public static void main(String[] args) throws Exception {
+ *       URL mojoURL = ExampleApp.class.getResource("/com/company/mojo.zip");
+ *       MojoReaderBackend reader = MojoReaderBackendFactory.createReaderBackend(mojoURL, CachingStrategy.MEMORY);
+ *       MojoModel model = ModelMojoReader.readFrom(reader);
+ *       EasyPredictModelWrapper modelWrapper = new EasyPredictModelWrapper(model);
+ *       RowData testRow = new RowData();
+ *         for (int i = 0; i < args.length; i++)
+ *           testRow.put("C"+i, Double.valueOf(args[i]));
+ *       RegressionModelPrediction prediction = (RegressionModelPrediction) modelWrapper.predict(testRow);
+ *       System.out.println("Prediction: " + prediction.value);
+ *     }
+ *   }
+ * }
+ * </pre>
+ */
+public class MojoReaderBackendFactory {
+
+  public enum CachingStrategy { MEMORY, DISK }
+
+  public static MojoReaderBackend createReaderBackend(String filename) throws IOException {
+    return createReaderBackend(new File(filename));
+  }
+
+  public static MojoReaderBackend createReaderBackend(File file) throws IOException {
+    if (file.isFile())
+      return new ZipfileMojoReaderBackend(file.getPath());
+    else if (file.isDirectory())
+      return new FolderMojoReaderBackend(file.getPath());
+    else
+      throw new IOException("Invalid file specification: " + file);
+  }
+
+  public static MojoReaderBackend createReaderBackend(URL url, CachingStrategy cachingStrategy) throws IOException {
+    try (InputStream is = url.openStream()) {
+      return createReaderBackend(is, cachingStrategy);
+    }
+  }
+
+  public static MojoReaderBackend createReaderBackend(InputStream inputStream, CachingStrategy cachingStrategy) throws IOException {
+    switch (cachingStrategy) {
+      case MEMORY:
+        return createInMemoryReaderBackend(inputStream);
+      case DISK:
+        return createTempFileReaderBackend(inputStream);
+    }
+    throw new IllegalStateException("Unexpected caching strategy: " + cachingStrategy);
+  }
+
+  private static MojoReaderBackend createInMemoryReaderBackend(InputStream inputStream) throws IOException {
+    HashMap<String, byte[]> content = new HashMap<>();
+    ZipInputStream zis = new ZipInputStream(inputStream);
+    try {
+      ZipEntry entry;
+      while ((entry = zis.getNextEntry()) != null) {
+        if (entry.getSize() > Integer.MAX_VALUE)
+          throw new IOException("File too large: " + entry.getName());
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        copyStream(zis, os);
+        content.put(entry.getName(), os.toByteArray());
+      }
+    } finally {
+      zis.close();
+    }
+    return new InMemoryMojoReaderBackend(content);
+  }
+
+  private static MojoReaderBackend createTempFileReaderBackend(InputStream inputStream) throws IOException {
+    File tmpFile = File.createTempFile("h2o-mojo", ".zip");
```

Comment:
should this have randomized name, in case 2+ processes are running in parallel ?